### PR TITLE
Bug 1704874: Copy Registry CA to Build Pods

### PR DESCRIPTION
**Pull request by adambkaplan**
With openshift/cluster-openshift-controller-manager-operator#96,
the openshift controller manager will be able to read the internal registry's
certificate authority from a central ConfigMap.
Updating build controller to copy the registry CA into each build's CA ConfigMap.

### CODE REVIEW

**Comment by adambkaplan** [comment]
/hold

Needs openshift/cluster-openshift-controller-manager-operator#96 to land first.

**Comment by adambkaplan** [comment]
/cc @deads2k 

ptal at the mutex commits - I intend to squash these separately from the changes to read the registry's CA.

File: pkg/build/controller/build/build_controller.go [hunk 1]
```diff
@@ -157,22 +157,24 @@ type BuildController struct {
...
-	secretStore                   v1lister.SecretLister
-	podStore                      v1lister.PodLister
-	imageStreamStore              imagev1lister.ImageStreamLister
-	openShiftConfigConfigMapStore v1lister.ConfigMapLister
+	buildStore                      buildv1lister.BuildLister
+	secretStore                     v1lister.SecretLister
+	podStore                        v1lister.PodLister
+	imageStreamStore                imagev1lister.ImageStreamLister
+	openShiftConfigConfigMapStore   v1lister.ConfigMapLister
+	controllerManagerConfigMapStore v1lister.ConfigMapLister
```
Comment (bparees) [inline pkg/build/controller/build/build_controller.go]: this is going to conflict with https://github.com/openshift/origin/pull/22734, can you rebase on top of it?

File: pkg/build/controller/build/build_controller.go [hunk 2]
```diff
@@ -971,13 +1042,12 @@ func (bc *BuildController) createBuildPod(build *buildv1.Build) (*buildUpdate, e
...
-	if len(bc.additionalTrustedCAData) > 0 {
-		includeAdditionalCA = true
-	}
+	// Get a copy of the additional trusted CAs
+	// We want to wire the same copy of the CA data through to avoid data races.
+	additionalCas := bc.additionalTrustedCAs()
+
 	// Create the build pod spec
-	buildPod, err := bc.createPodSpec(build, includeAdditionalCA)
+	buildPod, err := bc.createPodSpec(build, additionalCas)
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller.go]: s/additionalCas/additionalCAs

File: pkg/build/controller/build/build_controller.go [hunk 3]
```diff
@@ -1038,7 +1108,7 @@ func (bc *BuildController) createBuildPod(build *buildv1.Build) (*buildUpdate, e
 		}
 		if !hasCAMap {
 			// Create the CA ConfigMap to mount certificate authorities to the existing build pod
-			update, err = bc.createBuildCAConfigMap(build, existingPod, update)
+			update, err = bc.createBuildCAConfigMap(build, existingPod, update, additionalCas)
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller.go]: s/additionalCas/additionalCAs

File: pkg/build/controller/build/build_controller.go [hunk 4]
```diff
@@ -1058,7 +1128,7 @@ func (bc *BuildController) createBuildPod(build *buildv1.Build) (*buildUpdate, e
 	} else {
 		klog.V(4).Infof("Created pod %s/%s for build %s", build.Namespace, buildPod.Name, buildDesc(build))
 		// Create the CA ConfigMap to mount certificate authorities to the build pod
-		update, err = bc.createBuildCAConfigMap(build, pod, update)
+		update, err = bc.createBuildCAConfigMap(build, pod, update, additionalCas)
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller.go]: s/additionalCas/additionalCAs

File: pkg/build/controller/build/build_controller.go [hunk 5]
```diff
@@ -285,11 +290,76 @@ func NewBuildController(params *BuildControllerParams) *BuildController {
...
+		return caData
+	}
+	for k, v := range bc.additionalTrustedCAData {
+		caData[k] = v
+	}
+	return caData
+}
+
+func (bc *BuildController) setAdditionalTrustedCas(value map[string]string) {
+	bc.configLock.Lock()
```
Comment (bparees) [inline pkg/build/controller/build/build_controller.go]: aren't you only calling this once, during startup?  why do we need locks here and on the read above?

Comment (bparees) [reply pkg/build/controller/build/build_controller.go]: nm, i guess you're refreshing this data on every build?

Comment (bparees) [reply pkg/build/controller/build/build_controller.go]: or at least when the config object changes?

File: pkg/build/controller/build/build_controller.go [hunk 6]
```diff
@@ -1534,8 +1604,14 @@ func (bc *BuildController) handleBuildConfigError(err error, key interface{}) {
 }
 
 // createBuildCAConfigMap creates a ConfigMap containing certificate authorities used by the build pod.
-func (bc *BuildController) createBuildCAConfigMap(build *buildv1.Build, buildPod *corev1.Pod, update *buildUpdate) (*buildUpdate, error) {
-	configMapSpec := bc.createBuildCAConfigMapSpec(build, buildPod)
+func (bc *BuildController) createBuildCAConfigMap(build *buildv1.Build, buildPod *corev1.Pod, update *buildUpdate, additionalCas map[string]string) (*buildUpdate, error) {
+	configMapSpec, err := bc.createBuildCAConfigMapSpec(build, buildPod, additionalCas)
+	if err != nil {
+		bc.recorder.Eventf(build, corev1.EventTypeWarning, "FailedCreate", "Error creating build certificate authority configMap: %v", err)
+		update.setReason("CannotCreateCAConfigMap")
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller.go]: Should this be a const?

File: pkg/build/controller/build/build_controller.go [hunk 7]
```diff
@@ -1694,9 +1780,9 @@ func (bc *BuildController) readClusterImageConfig() []error {
 		configErrs = append(configErrs, err)
 		return configErrs
 	} else if imageConfig == nil {
-		bc.additionalTrustedCAData = nil
-		bc.registryConfData = ""
-		bc.signaturePolicyData = ""
+		bc.setAdditionalTrustedCas(nil)
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller.go]: s/setAdditionalTrustedCas/setAdditionalTrustedCAs

File: pkg/build/controller/build/build_controller.go [hunk 8]
```diff
@@ -1711,26 +1797,27 @@ func (bc *BuildController) readClusterImageConfig() []error {
 	if err != nil {
 		configErrs = append(configErrs, err)
 	} else {
-		bc.additionalTrustedCAData = additionalCAs
+		bc.setAdditionalTrustedCas(additionalCAs)
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller.go]: s/setAdditionalTrustedCas/setAdditionalTrustedCAs

File: pkg/build/controller/build/build_controller_test.go [hunk 9]
```diff
@@ -1201,38 +1217,27 @@ func TestCreateBuildCAConfigMap(t *testing.T) {
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
 			bc := newFakeBuildController(nil, nil, nil, nil, nil)
-			bc.additionalTrustedCAData = tc.addlCAData
+			bc.setAdditionalTrustedCas(tc.addlCAData)
```
Comment (coreydaley) [inline pkg/build/controller/build/build_controller_test.go]: s/Cas/CAs

File: pkg/build/controller/build/build_controller.go (hunk 5 above)
Comment (adambkaplan) [reply pkg/build/controller/build/build_controller.go]: No - we call this every sync for the cluster config.

File: pkg/build/controller/build/build_controller.go (hunk 1 above)
Comment (adambkaplan) [reply pkg/build/controller/build/build_controller.go]: will do.

File: pkg/build/controller/build/build_controller.go (hunk 5 above)
Comment (adambkaplan) [reply pkg/build/controller/build/build_controller.go]: We also run sync every time a ConfigMap in the `openshift-config` namespace changes.

File: pkg/build/controller/build/build_controller.go (hunk 6 above)
Comment (adambkaplan) [reply pkg/build/controller/build/build_controller.go]: not as likely now that creating the spec doesn't error.

**Comment by adambkaplan** [comment]
/hold cancel

ocm-o update has landed

File: pkg/build/controller/build/build_controller.go [hunk 10]
```diff
@@ -1551,27 +1627,39 @@ func (bc *BuildController) createBuildCAConfigMap(build *buildv1.Build, buildPod
...
 
-	if cm.Data == nil {
-		cm.Data = make(map[string]string)
+	// Fetch the registry's certificate authority stored in the openshift-service-ca ConfigMap
+	registryCAMap, err := bc.controllerManagerConfigMapStore.ConfigMaps("openshift-controller-manager").Get("openshift-service-ca")
+	// If we can't find the CA for the registry, pull from/push to the registry within a build will fail
+	// Logging warnings/errors because the origin/integration tests fail if we raise an error
+	if errors.IsNotFound(err) {
+		klog.V(2).Infof("WARNING - certificate authority for the internal registry could not be found. Image pulls/pushes to the internal registry within build %s/%s will fail.", build.Namespace, build.Name)
+		return cm
```
Comment (adambkaplan) [inline pkg/build/controller/build/build_controller.go]: @deads2k another case where the integration test breaks us. I'd rather fail the build here than wait for a `PullBuilderImage` error at execution time.

**Comment by adambkaplan** [comment]
@coreydaley @bparees @deads2k squashed & updated - ptal

File: pkg/build/controller/build/build_controller.go (hunk 10 above)
Comment (bparees) [reply pkg/build/controller/build/build_controller.go]: in a followup, make this an actual warning. (klog.Warningf)

File: pkg/build/controller/build/build_controller.go [hunk 11]
```diff
@@ -1551,27 +1627,39 @@ func (bc *BuildController) createBuildCAConfigMap(build *buildv1.Build, buildPod
...
 	}
-
+	if err != nil {
+		klog.V(1).Infof("ERROR - failed to read the internal registry's certificate authority. Image pulls/pushes to the internal registry within build %s/%s will fail. Error: %v", build.Namespace, build.Name, err)
+		return cm
+	}
+	if registryCAMap == nil || len(registryCAMap.Data) == 0 {
+		klog.V(2).Infof("WARNING - certificate authority for the internal registry not available. Image pulls/pushes to the internal registry within build %s/%s will fail.", build.Namespace, build.Name)
```
Comment (bparees) [inline pkg/build/controller/build/build_controller.go]: followup: actual warning

File: pkg/build/controller/build/build_controller.go [hunk 12]
```diff
@@ -1551,27 +1627,39 @@ func (bc *BuildController) createBuildCAConfigMap(build *buildv1.Build, buildPod
...
+	registryCAMap, err := bc.controllerManagerConfigMapStore.ConfigMaps("openshift-controller-manager").Get("openshift-service-ca")
+	// If we can't find the CA for the registry, pull from/push to the registry within a build will fail
+	// Logging warnings/errors because the origin/integration tests fail if we raise an error
+	if errors.IsNotFound(err) {
+		klog.V(2).Infof("WARNING - certificate authority for the internal registry could not be found. Image pulls/pushes to the internal registry within build %s/%s will fail.", build.Namespace, build.Name)
+		return cm
 	}
-
+	if err != nil {
+		klog.V(1).Infof("ERROR - failed to read the internal registry's certificate authority. Image pulls/pushes to the internal registry within build %s/%s will fail. Error: %v", build.Namespace, build.Name, err)
```
Comment (bparees) [inline pkg/build/controller/build/build_controller.go]: follow up: klog.Errorf

File: pkg/build/controller/build/build_controller.go [hunk 13]
```diff
@@ -1551,27 +1627,39 @@ func (bc *BuildController) createBuildCAConfigMap(build *buildv1.Build, buildPod
...
+		return cm
+	}
+	if registryCAMap == nil || len(registryCAMap.Data) == 0 {
+		klog.V(2).Infof("WARNING - certificate authority for the internal registry not available. Image pulls/pushes to the internal registry within build %s/%s will fail.", build.Namespace, build.Name)
+		return cm
+	}
+	registryCAData, exists := registryCAMap.Data[buildutil.ServiceCAKey]
+	if !exists {
+		klog.V(2).Infof("WARNING - certificate authority for the internal registry is missing. Image pulls/pushes to the internal registry within build %s/%s will fail.", build.Namespace, build.Name)
```
Comment (bparees) [inline pkg/build/controller/build/build_controller.go]: follow up: warningf

**Pull request merged**